app/javascript/components/widget-display.jsx

```diff
@@ -0,0 +1,216 @@
+import React, { useRef, useEffect, useReducer } from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+import { mathMod } from 'ramda';
+import { DisconnectedMessage, LoadingMessage } from '@messages/message';
+import FullPanel from '@components/full-panel';
+import SidePanel from '@components/side-panel';
+import Twitter from '@widgets/twitter';
+import Numbers from '@widgets/numbers';
+import Weather from '@widgets/weather';
+import Traffic from '@widgets/traffic';
+
+const Wrapper = styled.div`
+  outline: none;
+`;
+
+const widgetElements = {
+  Weather,
+  Twitter,
+  Numbers,
+  Traffic,
+};
+
+const prefetchDelay = 1500;
+let lastPrefetchTime = 0;
+
+const stateSwitch = (state, { type, changeValue }) => {
+  const { timerValue, pausedTime } = state;
+  switch (type) {
+    case 'setPausedTime':
+      return {
+        timerValue,
+        pausedTime: changeValue,
+      };
+    case 'extendTimer':
+      return {
+        timerValue: timerValue + changeValue,
+        pausedTime: null,
+      };
+    case 'resetTimerValue':
+      return {
+        timerValue: changeValue,
+        pausedTime,
+      };
+    default:
+      return state;
+  }
+};
+
+export const WidgetDisplay = ({
+  requestPrefetch,
+  requestWidget,
+  cityName,
+  widgets,
+}) => {
+  const { byIdOrFirst: current, enabled: enabledWidgets } = widgets;
+
+  const WidgetElement = widgetElements[current.name].Panel;
+
+  const index = enabledWidgets.findIndex(w => w.id === current.id);
+
+  const prevWidgetId =
+    enabledWidgets[mathMod(index - 1, enabledWidgets.length)].id;
+  const nextWidgetId =
+    enabledWidgets[mathMod(index + 1, enabledWidgets.length)].id;
```

Comment:
Thanks for taking the time to unravel this. Sometimes when we're working through a refactor, we discover other improvements that can be made in other sections of the application. In this case, it looks like we can't extract a `switchToNextWidget` and `switchToPrevWidget` callbacks because prev/next depend on the data from current widget state passed to the graphql query. A potential improvement is to expose `nextWidget(currWidget: Int)` and `prevWidget(currWidget: Int)` in the API itself. I'll make a ticket for this.